openedx/core/djangoapps/notifications/email/utils.py

```diff
@@ -243,3 +270,93 @@ def filter_notification_with_email_enabled_preferences(notifications, preference
         if notification.notification_type in enabled_course_prefs[notification.course_id]:
             filtered_notifications.append(notification)
     return filtered_notifications
+
+
+def encrypt_string(string):
+    """
+    Encrypts input string
+    """
+    return UsernameCipher.encrypt(string)
+
+
+def decrypt_string(string):
+    """
+    Decrypts input string
+    """
+    return UsernameCipher.decrypt(string).decode()
+
+
+def encrypt_object(obj):
+    """
+    Returns hashed string of object
+    """
+    string = json.dumps(obj)
+    return encrypt_string(string)
+
+
+def decrypt_object(string):
+    """
+    Decrypts input string and returns an object
+    """
+    decoded = decrypt_string(string)
+    return json.loads(decoded)
+
+
+def update_user_preferences_from_patch(encrypted_username, encrypted_patch):
+    """
+    Decrypt username and patch and updates user preferences
+    """
+    username = decrypt_string(encrypted_username)
+    patch = decrypt_object(encrypted_patch)
+
+    app_value = patch.get("app_name")
+    type_value = patch.get("notification_type")
+    channel_value = patch.get("channel")
+    pref_value = bool(patch.get("value", False))
+
+    kwargs = {'user__username': username}
+    if 'course_id' in patch.keys():
+        kwargs['course_id'] = patch['course_id']
+
+    def is_name_match(name, param_name):
+        """
+        Name is match if strings are equal or param_name is None
+        """
+        return True if param_name is None else name == param_name
+
+    def is_editable(app_name, notification_type, channel):
+        """
+        Returns if notification type channel is editable
+        """
+        if notification_type == 'core':
+            return channel not in COURSE_NOTIFICATION_APPS[app_name]['non_editable']
+        return channel not in COURSE_NOTIFICATION_TYPES[notification_type]['non_editable']
+
+    def get_default_cadence_value(app_name, notification_type):
+        """
+        Returns default email cadence value
+        """
+        if notification_type == 'core':
+            return COURSE_NOTIFICATION_APPS[app_name]['core_email_cadence']
+        return COURSE_NOTIFICATION_TYPES[notification_type]['email_cadence']
+
+    preferences = CourseNotificationPreference.objects.filter(**kwargs)
+    # pylint: disable=too-many-nested-blocks
+    for preference in preferences:
```

Comment:
It would be nice if this loop nesting could be flattened.

Reply:
For readability and performance, I think the best way for this loop is to remain nested.